src/runtime/simulator.cc

```diff
@@ -39,29 +39,841 @@ bool ParallelConfig::is_data_parallel() const
+  }
+
+  // Create inter node comm devices
+  for (int i = 0; i < num_nodes; i++) {
+    for (int j = 0; j < num_nodes; j++) {
+      if (i != j) {
+        int device_id = i * num_nodes + j;
+        std::string nic_name = "NIC " + std::to_string(device_id);
+        ids_to_inter_node_comm_device[device_id] = new Comm_device(nic_name, Comm_device::NIC_OUT_COMM, -1, -1, device_id, 0, inter_node_bandwidth);
+      }
+    }
+  }
+}
+
+int MachineModel_old::get_version()
+{
+  return version;
+}
+
+Comp_device *MachineModel_old::get_gpu(int device_id) 
+{
+  assert(id_to_gpu.find(device_id) != id_to_gpu.end());
+  return id_to_gpu[device_id];
+}
+
+Mem_device *MachineModel_old::get_gpu_fb_mem(int device_id) 
+{
+  assert(id_to_gpu_fb_mem.find(device_id) != id_to_gpu_fb_mem.end());
+  return id_to_gpu_fb_mem[device_id];
+}
+
+int MachineModel_old::get_num_gpus()
+{
+  return num_gpus;
+}
+
+float MachineModel_old::get_intra_node_gpu_bandwidth()
+{
+  return inter_gpu_bandwidth;
+}
+
+float MachineModel_old::get_inter_node_gpu_bandwidth()
+{
+  return inter_node_bandwidth;
+}
+
+
+std::vector<Comm_device *> MachineModel_old::get_comm_path(Mem_device *src_mem, Mem_device *tar_mem)
 {
-  printf("Device: node_id(%d) gpu_id(%d) capacity(%zu)\n", node_id, gpu_id, capacity);
-  assert(type == DEVICE_GPU);
+    std::vector<Comm_device *> ret;
+    // on the same memory
+    if (src_mem->mem_type == tar_mem->mem_type and src_mem->device_id == tar_mem->device_id) {
+        return ret;
+    }
+    if (src_mem->mem_type == Mem_device::SYSTEM_MEM and tar_mem->mem_type == Mem_device::SYSTEM_MEM) {
+        if (src_mem->node_id == tar_mem->node_id) {
+            return ret;
+        }
+        else {
+            int device_id = src_mem->node_id * num_nodes + tar_mem->node_id;
+            ret.emplace_back(ids_to_inter_node_comm_device[device_id]);
+        }
+    }
+    else if (src_mem->mem_type == Mem_device::GPU_FB_MEM and tar_mem->mem_type == Mem_device::GPU_FB_MEM) {
+        if (src_mem->node_id == tar_mem->node_id) {
+            int device_id = src_mem->device_id * num_gpus + tar_mem->device_id;
+            ret.emplace_back(ids_to_inter_gpu_comm_device[device_id]);
+        }
+        else {
+            ret.emplace_back(id_to_gputodram_comm_device[src_mem->device_id]);
+            int device_id = src_mem->node_id * num_nodes + tar_mem->node_id;
+            ret.emplace_back(ids_to_inter_node_comm_device[device_id]);
+            ret.emplace_back(id_to_dramtogpu_comm_device[tar_mem->device_id]);
+        }
+    }
+    else if (src_mem->mem_type == Mem_device::SYSTEM_MEM and tar_mem->mem_type == Mem_device::GPU_FB_MEM) {
+        if (src_mem->node_id == tar_mem->node_id) {
+            ret.emplace_back(id_to_dramtogpu_comm_device[tar_mem->device_id]);
+        }
+        else {
+            int device_id = src_mem->node_id * num_nodes + tar_mem->node_id;
```

Comment:
Can we replace a bunch of these places where we do the hashing explicitly with instead creating structs and providing them `std::hash` instances? The inline hashing code is bug-prone and makes things hard to read

Reply:
Make sense. When implementing the simple machine model, I tried to make it be the same as the original code, so the default behavior of Flexflow wouldn't be changed.